From: Jan Kara <jack@suse.cz>
References: fate#302681
Subject: ocfs2: Fix mount cleanup after quota failure
Patch-mainline: 2.6.29?

When we fail to initialize quotas during mount, the kernel oopses because
we tried to clean-up superblock twice.

Signed-off-by: Jan Kara <jack@suse.cz>

Index: linux-2.6.27/fs/ocfs2/super.c
===================================================================
--- linux-2.6.27.orig/fs/ocfs2/super.c	2008-10-31 07:59:08.000000000 +0100
+++ linux-2.6.27/fs/ocfs2/super.c	2008-10-31 08:18:35.000000000 +0100
@@ -972,8 +972,13 @@
 	if (!(sb->s_flags & MS_RDONLY)) {
 		status = ocfs2_enable_quotas(osb);
 		if (status < 0) {
+			/* We have to err-out specially here because
+			 * s_root is already set */
 			mlog_errno(status);
-			goto read_super_error;
+			atomic_set(&osb->vol_state, VOLUME_DISABLED);
+			wake_up(&osb->osb_mount_event);
+			mlog_exit(status);
+			return status;
 		}
 	}
 
